### ISSUE TITLE
Temporarily give up ownership of UserData struct

### ISSUE THREAD

**Issue by TheSilvus** [amethyst/rlua#146]
I want to be able to own a struct in Rust but be able to access it from within Lua.

This either requires the use of `Rc`/`Arc` (and along with that interior mutability) which I would like to avoid. The other way I could think of was temporarily give up ownership (or a mutable reference) whenever calling Lua code. I have tried this in two different ways but have been unable to get a satisfactory result:

1. I tried to use `Scope::create_nonstatic_userdata`. I created another container struct for a mutable reference to the thing I want to access and implemented `UserData` on the container. Using `create_nonstatic_userdata` prevents me from accessing that reference inside methods from `add_methods`, correct? I found the docs a little unclear here, they don't feel quite applicable to the methods they talk about.

2. I used `Scope::create_static_userdata` instead and gave up ownership of my struct. _Is there a way to get ownership back?_ I have been unable to find a method for this in the docs.

There might have been something I got fundamentally wrong here - I am neither a Rust nor a Lua expert. Is there a way to achieve what I want?

**Comment by kyren** [comment]
> I tried to use Scope::create_nonstatic_userdata. I created another container struct for a mutable reference to the thing I want to access and implemented UserData on the container. Using create_nonstatic_userdata prevents me from accessing that reference inside methods from add_methods, correct? I found the docs a little unclear here, they don't feel quite applicable to the methods they talk about.

You should have full access to the userdata inside methods added with `UserDataMethods::add_*method*` methods, so `add_method`, `add_method_mut`, `add_meta_method`, and `add_meta_method_mut`.  You can add the "function" version of those as well, but they won't be very helpful since you can't get the data back out of an `AnyUserData`.

The issues here are subtle and hard to explain, I'm sorry if the docs don't explain it very well.  Let me try to explain this a bit more fully...

Normally, userdata types in `rlua` must be 'static, and such userdata have metatables that are constructed once per type and put into a special table.  This makes constructing userdata of those much faster, when you create multiple userdata instances of the same 'static type, they all can share the same metatable.  The key for this metatable is the type's `TypeId`, which we can get because the type must be 'static.

This also allows you to check the type of such userdata by checking that the metatables are what you expect them to be.  If the metatables match, the type must be correct and then you can get a reference to the stored userdata out of an `AnyUserData` handle.  This is almost exactly the same thing as rust's `Any` trait and `Any::downcast_ref`, effectively you're just checking that the `TypeId` of two types match before doing a pointer conversion.

You'll notice though that the rust `Any` trait has a 'static bound on the type, and there's a very good reason for this.  You can only currently get a `TypeId` in rust for a 'static type because there is no mechanism in rust to distinguish between different *lifetimes* of different types at runtime.  I don't know the current status, but I believe at one point there was talk or an RFC for removing the 'static limitation from TypeId, but AIUI even with that limitation removed, the TypeId would still not reflect *lifetime* differences between types.  In other words, some `NonStatic<'a>` would have the same TypeId for all values of 'a.

I'm not 100% sure on this next part, I'm not a rust compiler dev, but I believe at some point the rust compiler "forgets" lifetime information, and there's a desire to keep the current situation where lifetimes cannot affect runtime behavior.  I also think this comes up when talking about specialization, such as the desire that you should *not* be able to specialize on different lifetimes, because that would be a way to affect runtime behavior with lifetimes.  Like I said I might be wrong about the details here, I haven't researched this in a while, but I don't think it's likely that we'll see "TypeId with lifetimes included" in the near future, or maybe ever.

So with that background information, how does the special "non-'static" userdata work?  It works entirely differently from other userdata types because it constructs a separate metatable on each call to `create_nonstatic_userdata`.  It is as though every call to `create_nonstatic_userdata` has a *unique* type, which is sort of true because each call will potentially have a different lifetime and there is no way to *know* which lifetime that is at runtime.

This means that anything that works by comparing `TypeId`s (or comparing metatables by identity, which is the same thing) just won't work.  This is the reason for the scare comments about the "function" type userdata methods, you'll get an `AnyUserData` handle that you won't be able to do anything with.

> I used Scope::create_static_userdata instead and gave up ownership of my struct. Is there a way to get ownership back? I have been unable to find a method for this in the docs.

The simple answer here is that the way to take ownership *from* Lua right now is just to have an `Option` and set it to `None`.  Lua has no concept of lifetimes, so you can't really take ownership of a value from Lua safely because there's no way to know whether Lua has a reference to it somewhere.

The more complex answer is that that's *kind of* a lie, there IS a way to actually take the value directly out of the userdata pointer inside Lua, I just don't expose it to the user yet.  the "scope" system has to "invalidate" userdata when the scope ends, so there is a process where the userdata can actually be moved out of the userdata pointer and the metatable replaced with a special "invalidated" metatable that just errors.  I *could* expose this to the user of `rlua` so that they could do the same thing without needing a special custom layer that involves an `Option`, I'll think about adding that.

**Comment by TheSilvus** [comment]
Thank you so much for your detailed answer! I think I got everything (most) of what you said.

Right now to use `Option` I'd need to wrap it into a container struct and implement `UserData` on that, correct? Would it be possible to provide a blanket implementation (not sure whether this is the right term) of `UserData` for `T: UserData` and error out when a method was called on `None`?

**Comment by kyren** [comment]
> Would it be possible to provide a blanket implementation (not sure whether this is the right term) of UserData for T: UserData and error out when a method was called on None?

I think that's basically equivalent to me exposing the "invalidation" mechanism to users, since that already exists and won't incur an extra option check.

**Comment by kyren** [comment]
I looked into quickly exposing the ability for users to manually "invalidate" userdata, and that might not be a great thing to try to do quickly 😛 

I think it would require more patience than I have right now, as the only things that actually invalidate userdata are `__gc` methods themselves and the "scope" system, but the "scope" system assumes that the userdata *has not* already been invalidated, and relies on holding references to the userdata to prevent destruction.

I'd have to add more checks to enable userdata to be manually invalidated at any time, and I don't think I'm prepared to do that right now.  In fact, I'm not convinced that the scope system is doing this properly, and I think it might actually be a good idea to make the scope system NOT use the same invalidation mechanism as the garbage collectors, but I'm also not quite prepared to do that right now either 😭

I've also found a potential corner case bug just by looking at it at all.  `rlua` is like a jar of spiders, nothing good can come from opening the jar even in the best case.
